text/0000-orderer-pruning-and-snapshot.md

```diff
@@ -0,0 +1,272 @@
+1. Deprecate the existing REST API endpoint **/v1/participation/channels** and discourage its use with the plan to drop it completely in versions >= 2.5
+2. Introduce new REST API endpoint **/v1/mgmt/channels** and keep all existing functionality of the channel participation API except "join by join-block", that is keep "remove channel" and "list channel(s)" APIs.
+3. Rename "Channel participation API Configuration" section in the OSN YAML config file to "Channel management API Configuration", rename accordingly all the config settings
+4. Add three new snapshot operations to the REST API definitions of **/v1/mgmt/channels** REST API
+5. Add support of three new snapshot operations to the CLI tool *osnadmin*, specifically to the command *osnadmin channel* and the new command *osnadmin snapshot*
+6. Deprecate *osnadmin channel join* command. This will be the only command using deprecated **/v1/participation/channels**, all the rest of the commands including inherited *osnadmin channel remove* and *osnadmin channel list* must use internally new **/v1/mgmt/channels** endpoint.
+
+## Snapshot content
+
+We propose the OSN snapshot to follow the same logical structure as peer snapshot, that is to comprise a directory with a number of files in it, one of the files being a metadata JSON file with the checksums of other files and other metadata. But here the similarity ends, as the content of the files will be drastically different.
+Here is the content of the OSN snapshot:
+1. Snapshot version - to allow for the future extensions. Current version is 1.
+2. Channel ID.
+3. Last config block for the channel.
+4. lastBlockNum - the number of the last pruned block. This can be both behind the last config block and ahead of it.
+5. lastBlockHash - the hash of the last pruned block.
+6. prevBlockHash - technically not required, but this will make internal structure BlockchainInfo complete.
+
+Unlike peer snapshot the OSN snapshot completely ignores the state DB, private data configs, the list of TxIDs and commit hash - all of these are not used by orderers. So these two snapshots bear resemblance only on the conceptual level, being completely different in the content.
+
+## Implementation considerations
+
+When a channel is joined from a snapshot either by "channel joinbysnapshot" or by "channel prune" command, we have to save the following info on the block storage:
+1. Last config block from snapshot
+2. BootstrappingSnapshotInfo
+
+For BootstrappingSnapshotInfo persistence the implementation is already there, but for the last config block a new persistent unit of storage (a file in case of file-based ledger) should be created. Additionally all the procedures looking up the last config block should be modified to return the last config block from the snapshot in case the last config block is not found in the block ledger itself.
+
+## System channel co-existence
+
+The following APIs can work both with system channel and without it:
```

Comment:
I don't think we should mix snapshots and system channel. It's best to simply disable the feature in case we are using a system channel. 

You can't "use" the snapshot with a system channel anyway, so there is no point in listing it or creating snapshots.

Reply:
We included it here for two reasons:
1. It is technically possible, snapshot creation for a channel does not depend on the method of the channel creation (with or without system channel)
2. For the future expansion when snapshot will have more interesting content we could use snapshots from different orderers to verify that they are properly synchronized. This is one if the use cases for peer snapshots.

We understand that both arguments are weak, and if we need this in future we can easily allow snapshot creation in the presence of system channel anytime later.
So I will move all the "snapshot" APIs to the section "will fail in presence of system channel".

Reply:
Done